dot/sync/chain_sync.go

```diff
@@ -169,23 +176,48 @@ func newChainSync(cfg chainSyncConfig) *chainSync {
 	}
 }
 
+func (cs *chainSync) waitEnoughPeersAndTarget() <-chan struct{} {
+	resultCh := make(chan struct{})
+	go func() {
+		defer close(resultCh)
+		for {
+			select {
+			case <-resultCh:
+				return
+			default:
+			}
+
+			cs.workerPool.useConnectedPeers()
+			_, err := cs.getTarget()
+
+			totalAvailable := cs.workerPool.totalWorkers()
+			if totalAvailable >= uint(cs.minPeers) && err == nil {
+				return
+			}
+
+			// TODO: https://github.com/ChainSafe/gossamer/issues/3402
+			time.Sleep(100 * time.Millisecond)
+		}
+	}()
+
+	return resultCh
+}
+
 func (cs *chainSync) start() {
 	// since the default status from sync mode is syncMode(tip)
 	isSyncedGauge.Set(1)
 
+	cs.wg.Add(1)
+	go cs.pendingBlocks.run(cs.finalisedCh, cs.stopCh, &cs.wg)
+
 	// wait until we have a minimal workers in the sync worker pool
-	for {
-		totalAvailable := cs.workerPool.totalWorkers()
-		if totalAvailable >= uint(cs.minPeers) {
-			break
-		}
+	resultCh := cs.waitEnoughPeersAndTarget()
 
-		// TODO: https://github.com/ChainSafe/gossamer/issues/3402
-		time.Sleep(time.Second)
+	select {
+	case <-resultCh:
```

Comment:
why are we reading from this channel in two different spots? Here and on line 185?

Reply:
Answered here https://github.com/ChainSafe/gossamer/pull/3392#issuecomment-1788953984